apps/app_voicemail.c

```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
+{
+	const char *mailbox = a->argv[2];
+	const char *context = a->argv[3];
+	const char *from_folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	const char *to_folder = a->argv[6];
+	int ret =  vm_msg_move(mailbox, context, 1, from_folder, id, to_folder);
+	if (ret) {
+		ast_cli(a->fd, "Error moving message %s from mailbox %s@%s %s to %s\n",
+					id[0], mailbox, context, from_folder, to_folder);
+	}
+	else{
+		ast_cli(a->fd, "Succesfully moved message %s from mailbox %s@%s %s to %s\n",
+					id[0], mailbox, context, from_folder, to_folder);
+	}
+	return ret;
+}
+
+static int remove_message_from_mailbox(struct ast_cli_args *a)
+{
+	const char *mailbox = a->argv[2];
+	const char *context = a->argv[3];
+	const char *folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	int ret = vm_msg_remove(mailbox, context, 1, folder, id);
+	if (ret) {
+		ast_cli(a->fd, "Error removing message %s from mailbox %s@%s %s\n",
+					id[0], mailbox, context, folder);
+	}
+	else{
+		ast_cli(a->fd, "Succesfully removed message %s from mailbox %s@%s %s\n",
+					id[0], mailbox, context, folder);
+	}
+	return ret;
+}
+
+static char *complete_voicemail_show_mailbox(struct ast_cli_args *a)
+{
+	const char *word = a->word;
+	int pos = a->pos;
+	int state = a->n;
+	int which = 0;
+	int wordlen;
+	struct ast_vm_user *vmu;
+	const char *context = "", *mailbox = "";
+	char *ret = NULL;
+
+	/* 0 - voicemail; 1 - show; 2 - mailbox; 3 - <mailbox>; 4 - <context> */
+	if (pos == 3) {
+		wordlen = strlen(word);
+		AST_LIST_LOCK(&users);
+		AST_LIST_TRAVERSE(&users, vmu, list) {
+			if (!strncasecmp(word, vmu->mailbox , wordlen)) {
+				if (mailbox && strcmp(mailbox, vmu->mailbox) && ++which > state) {
+					ret = ast_strdup(vmu->mailbox);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				mailbox = vmu->mailbox;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	} else if (pos == 4) {
+		/* Only display contexts that match the user in pos 3 */
+		const char *box = a->argv[3];
+		wordlen = strlen(word);
+		AST_LIST_LOCK(&users);
+		AST_LIST_TRAVERSE(&users, vmu, list) {
+			if (!strncasecmp(word, vmu->context, wordlen) && !strcasecmp(box, vmu->mailbox)) {
+				if (context && strcmp(context, vmu->context) && ++which > state) {
+					ret = ast_strdup(vmu->context);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				context = vmu->context;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	}
+
+	return ret;
+}
+
+static char *handle_voicemail_show_mailbox(struct ast_cli_entry *e, int cmd, struct ast_cli_args *a)
+{
+	switch (cmd) {
+	case CLI_INIT:
+		e->command = "voicemail show mailbox";
+		e->usage =
+		"Usage: voicemail show mailbox <mailbox> <context>\n"
+		"       Show contents of mailbox <mailbox>@<context>\n";
+		return NULL;
+	case CLI_GENERATE:
+		return complete_voicemail_show_mailbox(a);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 5) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (show_messages_for_mailbox(a)) {
```

Comment:
This function always returns 0, so this warning is unreachable. And as is, this warning message isn't really helpful since it lacks details.

Reply:
changed the underlying behavior and to return the CLI error like the other commands.